### DISCUSSION TITLE
Message unacknowlegment but i can not consume it

### DISCUSSION THREAD

**Discussion by ThinhLe30** [apache/pulsar#22059] [Q&A] [answered]
According to Apache Pulsar official document, it said that 
`By default, Pulsar message brokers: immediately delete all messages that have been acknowledged by a consumer`

Here is my prob:
1. I set up docker-compose for Pulsar cluster:

```version: "3.1"
services:
  standalone:
    image: apachepulsar/pulsar:2.6.0
    ports:
      - 8088:8088
      - 6650:6650
    environment:
      PULSAR_MEM: " -Xms512m -Xmx512m -XX:MaxDirectMemorySize=1g"
    command: bin/pulsar standalone
```

2. I write a simple Pulsar Producer

```
client, err := pulsar.NewClient(pulsar.ClientOptions{
		URL: "pulsar://localhost:59138",
	})
	
	defer client.Close()
	
	producer, err := client.CreateProducer(pulsar.ProducerOptions{
		Topic: "my-topic",
	})
	
	_, err = producer.Send(context.Background(), &pulsar.ProducerMessage{
		Payload: []byte("hello"),

	})

	defer producer.Close()
	
	if err != nil {
		fmt.Println("Failed to publish message", err)
	} else {
		fmt.Println("Published message")
	}
```

3. I write a simple Pulsar Consumer

```
client, err := pulsar.NewClient(pulsar.ClientOptions{
		URL: "pulsar://localhost:59138",
	})
	
	defer client.Close()
	
	consumer, err := client.Subscribe(pulsar.ConsumerOptions{
			Topic:            "my-topic",
			SubscriptionName: "my-sub",
			Type:             pulsar.Shared,
		})
	
	defer consumer.Close()
	
	msg, err := consumer.Receive(context.Background())
		if err != nil {
			log.Fatal(err)
		}
	consumer.Ack(msg)
	fmt.Printf("Received message msgId: %#v -- content: '%s'\n",
				msg.ID(), string(msg.Payload()))
```
Both the code script following this library [Apache Pulsar Go Client Library](https://github.com/apache/pulsar-client-go)

And when I first run the producer and follow the consumer, it doesn't consume any message even though i have produced it. I had to run the consumer first, then the producer, i received the message successfully.
But i noticed that it happened only the first time when i started the docker container, it's weird and I'm new to Pulsar, can someone explain that to me :(( many thanks

**Comment by jak78** [comment, marked as answer]
Hello @ThinhLe30, and welcome to the community! The behavior you're experiencing might be caused by the default setting of the initial subscription position as `Latest`. This setting means that the subscription will only deliver messages that are received after the subscription is created.

To modify this behavior, you can opt for the `Earliest` option in [SubscriptionInitialPosition](https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar#SubscriptionInitialPosition) within the `ConsumerOptions` when creating the consumer. By selecting Earliest, the subscription will instead deliver the oldest messages that have not yet been consumed, starting from the earliest available.

Feel free to try the following:

```go
consumer, err := client.Subscribe(pulsar.ConsumerOptions{
			Topic:            "my-topic",
			SubscriptionName: "my-sub",
			Type:             pulsar.Shared,
			SubscriptionInitialPosition: pulsar.Earliest,
		})
```

**Reply by ThinhLe30** [reply]
Thanks a lot @jak78, wish you have a good day.

**Reply by jak78** [reply]
@ThinhLe30 I'm pleased to hear that this resolved your issue. By the way, I noticed that you're using Pulsar 2.6.0, which is no longer supported. I recommend upgrading to the most recent version for better stability and features:
- For long-term support, use version 3.0.2.
- For the latest features, consider version 3.2.0.

You can review the release policy to choose the version that best suits your needs: https://pulsar.apache.org/contribute/release-policy/

Best,
